master/pkg/schemas/expconf/historical.go

```diff
@@ -0,0 +1,175 @@
+		return bm, errors.Wrap(err, "historical bind mounts does not pass sanity checks")
+	}
+	if err = json.Unmarshal(bmByts, &bm); err != nil {
+		return bm, errors.Wrap(err, "unable to unmarshal bind mounts bytes")
+	}
+	bm = schemas.WithDefaults(bm).(BindMountsConfig)
+	if err = schemas.IsComplete(bm); err != nil {
+		return bm, errors.Wrap(err, "historical bind mounts is incomplete")
+	}
+	return bm, nil
+}
+
+func getEnvironmentVariables(raw map[string]interface{}) (EnvironmentVariablesMap, error) {
+	ev := EnvironmentVariablesMap{}
+
+	envOnly := raw["environment"]
+	if envOnly == nil {
+		// Empty environment.
+		return ev, nil
+	}
+
+	evOnly := envOnly.(map[string]interface{})["environment_variables"]
+	if evOnly == nil {
+		// Empty environment.environment_variables.
+		return ev, nil
+	}
+
+	evByts, err := json.Marshal(evOnly)
+	if err != nil {
+		return ev, errors.Wrap(err, "unable to remarshal environment variables as json")
+	}
+	ev = EnvironmentVariablesMap{}
+	if err = schemas.SaneBytes(&ev, evByts); err != nil {
+		return ev, errors.Wrap(err, "historical environment variables does not pass sanity checks")
+	}
+	if err = json.Unmarshal(evByts, &ev); err != nil {
+		return ev, errors.Wrap(err, "unable to unmarshal environment variables bytes")
+	}
+	ev = schemas.WithDefaults(ev).(EnvironmentVariablesMap)
+	if err = schemas.IsComplete(ev); err != nil {
+		return ev, errors.Wrap(err, "historical environment variables is incomplete")
+	}
+	return ev, nil
+}
+
+// ParseHistoricalConfigJSON parses bytes that represent an experiment config that was once valid
+// but might no longer be shimmable to the current version of ExperimentConfig.
+func ParseHistoricalConfigJSON(byts []byte) (HistoricalConfig, error) {
+	// Known difficulties (as of 0.15.5) that we are implicitly avoiding here:
+	// - Pre-remove-steps configs cannot be parsed as ExperimentConfig because several
+	//   fields changed, including but not limited to batches_per_step, searcher.max_length.
+	//   We are implicitly avoiding this by simply stripping all fields except checkpoint_storage.
+	// - There was a removal of adaptive and adaptive_simple.  As of 0.15.5, some of that is handled
+	//   by reading select parts of the searcher config in SQL queries.  Eventually that may get
+	//   handled here, in which case we need to be aware of that removal.
+	out := HistoricalConfig{}
+
+	raw := map[string]interface{}{}
+	if err := json.Unmarshal(byts, &raw); err != nil {
+		return out, errors.Wrap(err, "unable to unmarshal bytes as json at all")
+	}
+
+	cs, err := getCheckpointStorage(raw)
+	if err != nil {
+		return out, err
+	}
+	out.checkpointStorage = cs
+
+	bm, err := getBindMounts(raw)
+	if err != nil {
+		return out, err
+	}
+	out.bindMounts = bm
+
+	ev, err := getEnvironmentVariables(raw)
+	if err != nil {
+		return out, err
+	}
+	out.envvars = ev
+
+	return out, nil
+}
+
+// Scan implements the db.Scanner interface.
+func (h *HistoricalConfig) Scan(src interface{}) error {
```

Comment:
not really much of a difference, but i prefer querying bytes and putting this logic in the db rather than hiding db specific logic (impl scan) in expconf.

Reply:
Sure, I can do that.  The anonymous struct to invoke the Scan was a little weird anyway

Reply:
this is done now.